pkg/logging/limit.go

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package logging
+
+import (
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"golang.org/x/time/rate"
+)
+
+const messageKey = "msg"
+
+type occurrence struct {
+	lastSeen uint32
+	current  uint32
+}
+
+type rateLimitedLogger struct {
+	logger  log.Logger
+	limiter *rate.Limiter
+
+	m           sync.Mutex
+	occurrences map[string]*occurrence
+}
+
+// Log limits the amount of events logged and keep tracks of occurrences of messages.
+func (r *rateLimitedLogger) Log(keyvals ...interface{}) error {
+	var msg string
+	for i := range keyvals {
+		if i%2 == 0 && keyvals[i] == messageKey {
```

Comment:
This is something it would be awesome to avoid. buffering tons of memory in logger does not sounds like a good idea (: 

Having rate-limiting for single messages sounds better to me.